File: src/app/ui/prompt.zig
const std = @import("std");
const builtin = @import("builtin");
const zli = @import("zli");

const github = @import("../../core/github.zig");
const output = @import("output.zig");
const styles = @import("styles.zig");

const prompt_text = struct {
    const title = "Choose action updates";
    const controls_summary = "Move selection with arrows or j/k";
    const footer = "Up/Down or j/k move  <space> row  <f> file  <enter> apply  <a> all  <i> invert  <n> none  <q> cancel";
};

const posix = std.posix;

pub const Error = error{
    UnsupportedPlatform,
    NotATerminal,
    Canceled,
    Interrupted,
} || std.mem.Allocator.Error || std.posix.TermiosGetError || std.posix.TermiosSetError || std.posix.ReadError || std.Io.Writer.Error;

const Key = enum {
    up,
    down,
    scroll_up,
    scroll_down,
    toggle,
    toggle_all,
    toggle_file,
    invert,
    select_none,
    accept,
    cancel,
    ignore,
};

const Widths = struct {
    action: usize,
    job: usize,
    current: usize,
};

const visible_rows = 18;
const header_lines = 4;

const RawTerminal = struct {
    original: posix.termios,
    active: bool = false,

    fn enter() !RawTerminal {
        const original = try posix.tcgetattr(posix.STDIN_FILENO);
        var raw = original;

        raw.iflag.ICRNL = false;
        raw.iflag.IXON = false;
        raw.lflag.ECHO = false;
        raw.lflag.ICANON = false;
        raw.lflag.IEXTEN = false;
        raw.cc[@intFromEnum(posix.V.MIN)] = 0;
        raw.cc[@intFromEnum(posix.V.TIME)] = 1;

        try posix.tcsetattr(posix.STDIN_FILENO, .FLUSH, raw);
        return .{ .original = original, .active = true };
    }

    fn deinit(self: *RawTerminal) void {
        if (!self.active) return;
        posix.tcsetattr(posix.STDIN_FILENO, .FLUSH, self.original) catch {};
        self.active = false;
    }
};

pub fn selectUpdates(
    allocator: std.mem.Allocator,
    ctx: zli.CommandContext,
    candidates: []const github.Candidate,
) Error![]usize {
    if (builtin.os.tag == .windows) return error.UnsupportedPlatform;
    if (candidates.len == 0) return allocator.alloc(usize, 0);
    if (!try std.Io.File.stdin().isTty(ctx.io)) return error.NotATerminal;

    var selected = try allocator.alloc(bool, candidates.len);
    defer allocator.free(selected);
    @memset(selected, false);

    var terminal = RawTerminal.enter() catch |err| switch (err) {
        error.NotATerminal => return error.NotATerminal,
        else => return err,
    };
    defer terminal.deinit();

    try ctx.writer.print(styles.HIDE_CURSOR, .{});
    defer {
        ctx.writer.print(styles.SHOW_CURSOR, .{}) catch {};
        ctx.writer.flush() catch {};
    }

    var cursor: usize = 0;
    var scroll: usize = 0;
    var rendered_dynamic_lines: usize = 0;
    defer clearPrompt(ctx, header_lines + rendered_dynamic_lines) catch {};

    try renderHeader(ctx, candidates);

    while (true) {
        adjustScroll(&scroll, cursor, candidates.len);
        try clearPrompt(ctx, rendered_dynamic_lines);
        rendered_dynamic_lines = try renderDynamic(ctx, candidates, selected, cursor, scroll);

        switch (try readKey()) {
            .up => cursor = if (cursor == 0) candidates.len - 1 else cursor - 1,
            .down => cursor = if (cursor + 1 == candidates.len) 0 else cursor + 1,
            .scroll_up => scrollUp(&scroll, &cursor, candidates.len),
            .scroll_down => scrollDown(&scroll, &cursor, candidates.len),
            .toggle => selected[cursor] = !selected[cursor],
            .toggle_all => toggleAll(selected),
            .toggle_file => toggleFile(candidates, selected, candidates[cursor].file),
            .invert => invertSelected(selected),
            .select_none => @memset(selected, false),
            .accept => break,
            .cancel => return error.Canceled,
            .ignore => {},
        }
    }

    var indexes: std.ArrayList(usize) = .empty;
    errdefer indexes.deinit(allocator);

    for (selected, 0..) |is_selected, index| {
        if (is_selected) try indexes.append(allocator, index);
    }

    return indexes.toOwnedSlice(allocator);
}

fn renderHeader(ctx: zli.CommandContext, candidates: []const github.Candidate) Error!void {
    try ctx.writer.print("{s}{s}{s}\n\n", .{ styles.BOLD, prompt_text.title, styles.RESET });
    try ctx.writer.print("{s}•{s} {s}{d}{s} update candidates across {s}{d}{s} workflow files\n", .{
        styles.GREEN,
        styles.RESET,
        styles.YELLOW,
        candidates.len,
        styles.RESET,
        styles.YELLOW,
        workflowCount(candidates),
        styles.RESET,
    });
    try ctx.writer.print("{s}?{s} {s}{s}\n", .{
        styles.CYAN,
        styles.RESET,
        prompt_text.controls_summary,
        styles.RESET,
    });
    try ctx.writer.flush();
}

fn renderDynamic(
    ctx: zli.CommandContext,
    candidates: []const github.Candidate,
    selected: []const bool,
    cursor: usize,
    scroll: usize,
) Error!usize {
    var lines: usize = 0;

    const widths = Widths{
        .action = maxWidth(candidates, .action),
        .job = maxWidth(candidates, .job),
        .current = maxWidth(candidates, .current),
    };

    const end = @min(candidates.len, scroll + visible_rows);
    if (scroll > 0) {
        try ctx.writer.print("{s}↑ {d} more above{s}\n", .{ styles.BRIGHT_BLACK, scroll, styles.RESET });
        lines += 1;
    }

    var current_file: ?[]const u8 = null;
    for (candidates[scroll..end], scroll..) |candidate, index| {
        if (current_file == null or !std.mem.eql(u8, current_file.?, candidate.file)) {
            current_file = candidate.file;
            lines += try renderGroupHeader(ctx, candidate.file, widths);
        }

        try renderRow(ctx, candidate, selected[index], index == cursor, widths);
        lines += 1;
    }

    if (end < candidates.len) {
        try ctx.writer.print("{s}↓ {d} more below{s}\n", .{ styles.BRIGHT_BLACK, candidates.len - end, styles.RESET });
        lines += 1;
    }

    try ctx.writer.print("\n{s}{s}{s}\n", .{ styles.CYAN, prompt_text.footer, styles.RESET });
    lines += 2;
    try ctx.writer.flush();

    return lines;
}

fn renderGroupHeader(ctx: zli.CommandContext, file: []const u8, widths: Widths) Error!usize {
    try ctx.writer.print("\n{s}› {s}{s}\n", .{ styles.CYAN, file, styles.RESET });
    try ctx.writer.print("  {s}○ Action{s}", .{ styles.BRIGHT_BLACK, styles.RESET });
    try writePadding(ctx, trailingPadding(widths.action, "Action".len, 4));
    try ctx.writer.print("{s}Job{s}", .{ styles.BRIGHT_BLACK, styles.RESET });
    try writePadding(ctx, trailingPadding(widths.job, "Job".len, 4));
    try ctx.writer.print("{s}Current{s}", .{ styles.BRIGHT_BLACK, styles.RESET });
    try writePadding(ctx, trailingPadding(widths.current, "Current".len, 14));
    try ctx.writer.print("{s}› Target{s}\n", .{ styles.BRIGHT_BLACK, styles.RESET });

    return 3;
}

fn renderRow(
    ctx: zli.CommandContext,
    row_candidate: github.Candidate,
    is_selected: bool,
    is_cursor: bool,
    widths: Widths,
) Error!void {
    const row_style = if (is_cursor) styles.INVERSE else styles.DIM;
    const pointer = if (is_cursor) "›" else " ";
    const checkbox_color = if (is_selected) styles.GREEN else styles.BRIGHT_BLACK;
    const checkbox = if (is_selected) "●" else "○";
    const target_color = if (row_candidate.hasShaMismatch()) styles.YELLOW else if (row_candidate.isMajorUpdate()) styles.RED else styles.GREEN;
    const target = row_candidate.displayTarget();

    try ctx.writer.print("{s}{s}{s}  {s}{s}{s} {s}{s}{s}{s}", .{
        row_style,
        pointer,
        styles.RESET,
        checkbox_color,
        checkbox,
        styles.RESET,
        row_style,
        styles.BOLD,
        row_candidate.action,
        styles.RESET,
    });
    try writePadding(ctx, trailingPadding(widths.action, row_candidate.action.len, 4));

    try ctx.writer.print("{s}{s}{s}", .{ row_style, row_candidate.job, styles.RESET });
    try writePadding(ctx, trailingPadding(widths.job, row_candidate.job.len, 4));

    try ctx.writer.print("{s}{s}{s}", .{ styles.BOLD, row_candidate.current, styles.RESET });
    if (row_candidate.hasVersionComment()) {
        const mismatch_color = if (row_candidate.hasShaMismatch()) styles.RED else styles.BRIGHT_BLACK;
        try ctx.writer.print(" {s}({s}){s}", .{ mismatch_color, row_candidate.version_comment, styles.RESET });
    } else if (row_candidate.hasCurrentRef()) {
        try ctx.writer.print(" {s}({s}){s}", .{ styles.BRIGHT_BLACK, shortSha(row_candidate.current_ref), styles.RESET });
    }
    try writePadding(ctx, trailingPadding(widths.current, row_candidate.current.len, 4));

    try ctx.writer.print("{s}›{s}  {s}{s}{s}\n", .{
        styles.BOLD,
        styles.RESET,
        target_color,
        target,
        styles.RESET,
    });
}

fn adjustScroll(scroll: *usize, cursor: usize, count: usize) void {
    if (count <= visible_rows) {
        scroll.* = 0;
        return;
    }

    if (cursor < scroll.*) {
        scroll.* = cursor;
        return;
    }

    if (cursor >= scroll.* + visible_rows) {
        scroll.* = cursor + 1 - visible_rows;
    }
}

fn clearPrompt(ctx: zli.CommandContext, rendered_lines: usize) Error!void {
    if (rendered_lines == 0) return;

    try ctx.writer.print("\x1b[{d}F\x1b[J", .{rendered_lines});
    try ctx.writer.flush();
}

fn shortSha(sha: []const u8) []const u8 {
    return sha[0..@min(7, sha.len)];
}

fn readKey() Error!Key {
    var byte: [1]u8 = undefined;
    var read_len: usize = 0;
    while (read_len == 0) {
        read_len = try posix.read(posix.STDIN_FILENO, &byte);
    }

    return switch (byte[0]) {
        3 => error.Interrupted,
        4, 'q' => .cancel,
        27 => keyOrEscapeSequence(),
        '\r', '\n' => .accept,
        ' ', 'x' => .toggle,
        'a' => .toggle_all,
        'f' => .toggle_file,
        'i' => .invert,
        'n' => .select_none,
        'j' => .down,
        'k' => .up,
        else => .ignore,
    };
}

fn keyOrEscapeSequence() Error!Key {
    var seq: [2]u8 = undefined;
    const len = posix.read(posix.STDIN_FILENO, &seq) catch return .cancel;
    if (len < 2 or seq[0] != '[') return .cancel;

    return switch (seq[1]) {
        'A' => .up,
        'B' => .down,
        else => .ignore,
    };
}

fn scrollUp(scroll: *usize, cursor: *usize, count: usize) void {
    if (scroll.* == 0) return;
    scroll.* -= 1;

    if (cursor.* >= scroll.* + @min(visible_rows, count)) {
        cursor.* = scroll.* + @min(visible_rows, count) - 1;
    }
}

fn scrollDown(scroll: *usize, cursor: *usize, count: usize) void {
    if (count <= visible_rows) return;

    const max_scroll = count - visible_rows;
    if (scroll.* >= max_scroll) return;
    scroll.* += 1;

    if (cursor.* < scroll.*) {
        cursor.* = scroll.*;
    }
}

const WidthField = enum { action, job, current };

fn maxWidth(candidates: []const github.Candidate, field: WidthField) usize {
    var width: usize = 0;
    for (candidates) |candidate| {
        const len = switch (field) {
            .action => candidate.action.len,
            .job => candidate.job.len,
            .current => candidate.current.len,
        };
        width = @max(width, len);
    }
    return width;
}

fn trailingPadding(column_width: usize, value_len: usize, gap: usize) usize {
    return column_width -| value_len + gap;
}

fn workflowCount(candidates: []const github.Candidate) usize {
    var count: usize = 0;
    var last_file: ?[]const u8 = null;
    for (candidates) |candidate| {
        if (last_file == null or !std.mem.eql(u8, last_file.?, candidate.file)) {
            count += 1;
            last_file = candidate.file;
        }
    }
    return count;
}

fn invertSelected(selected: []bool) void {
    for (selected) |*item| {
        item.* = !item.*;
    }
}

fn toggleAll(selected: []bool) void {
    var all_selected = true;
    for (selected) |item| {
        if (!item) {
            all_selected = false;
            break;
        }
    }
    @memset(selected, !all_selected);
}

fn toggleFile(candidates: []const github.Candidate, selected: []bool, file: []const u8) void {
    var all_selected = true;
    for (candidates, 0..) |candidate, index| {
        if (!std.mem.eql(u8, candidate.file, file)) continue;
        if (!selected[index]) {
            all_selected = false;
            break;
        }
    }

    for (candidates, 0..) |candidate, index| {
        if (std.mem.eql(u8, candidate.file, file)) {
            selected[index] = !all_selected;
        }
    }
}

fn writePadding(ctx: zli.CommandContext, count: usize) Error!void {
    var i: usize = 0;
    while (i < count) : (i += 1) {
        try ctx.writer.writeByte(' ');
    }
}

test "trailing padding does not underflow for header labels wider than values" {
    try std.testing.expectEqual(@as(usize, 14), trailingPadding(2, "Current".len, 14));
    try std.testing.expectEqual(@as(usize, 4), trailingPadding(3, "Action".len, 4));
    try std.testing.expectEqual(@as(usize, 11), trailingPadding(10, "Job".len, 4));
}

File: src/ui/prompt.rs
use std::collections::HashSet;
use std::io::{self, IsTerminal, Stdout};

use crossterm::event::{self, Event, KeyCode, KeyEventKind, KeyModifiers};
use crossterm::execute;
use crossterm::terminal::{
    EnterAlternateScreen, LeaveAlternateScreen, disable_raw_mode, enable_raw_mode,
};
use ratatui::Terminal;
use ratatui::backend::CrosstermBackend;
use ratatui::layout::{Constraint, Layout};
use ratatui::style::{Color, Modifier, Style};
use ratatui::text::{Line, Span};
use ratatui::widgets::{Block, Borders, List, ListItem, ListState, Paragraph, Wrap};

use crate::model::ResolvedUpdate;

const FOOTER: &str =
    "Up/Down/j/k move  space row  f file  tab fold  enter apply  a all  i invert  n none  q cancel";
const VISIBLE_ROWS_HINT: usize = 12;

#[derive(Debug)]
pub enum Error {
    NotATerminal,
    Canceled,
    Interrupted,
    Io(io::Error),
}

impl std::fmt::Display for Error {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::NotATerminal => f.write_str("interactive selection requires a terminal"),
            Self::Canceled => f.write_str("selection canceled"),
            Self::Interrupted => f.write_str("selection interrupted"),
            Self::Io(err) => err.fmt(f),
        }
    }
}

impl std::error::Error for Error {}

impl From<io::Error> for Error {
    fn from(value: io::Error) -> Self {
        Self::Io(value)
    }
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
enum Key {
    Up,
    Down,
    Toggle,
    ToggleAll,
    ToggleFile,
    ToggleCollapse,
    Invert,
    SelectNone,
    Accept,
    Cancel,
    Ignore,
}

enum VisibleRow {
    FileHeader { file: String },
    Update { original_index: usize },
}

fn build_visible_rows(updates: &[ResolvedUpdate], collapsed: &HashSet<String>) -> Vec<VisibleRow> {
    let mut rows = Vec::new();
    let mut last_file: Option<&str> = None;

    for (index, update) in updates.iter().enumerate() {
        let file = update.file();
        let is_first = last_file != Some(file);

        if is_first {
            rows.push(VisibleRow::FileHeader {
                file: file.to_string(),
            });
            if !collapsed.contains(file) {
                rows.push(VisibleRow::Update {
                    original_index: index,
                });
            }
        } else if !collapsed.contains(file) {
            rows.push(VisibleRow::Update {
                original_index: index,
            });
        }

        last_file = Some(file);
    }

    rows
}

fn file_at_cursor(
    visible_rows: &[VisibleRow],
    cursor: usize,
    updates: &[ResolvedUpdate],
) -> String {
    if cursor >= visible_rows.len() {
        return updates[0].file().to_string();
    }
    match &visible_rows[cursor] {
        VisibleRow::FileHeader { file } => file.clone(),
        VisibleRow::Update { original_index } => updates[*original_index].file().to_string(),
    }
}

pub fn select_updates(updates: &[ResolvedUpdate]) -> Result<Vec<usize>, Error> {
    if updates.is_empty() {
        return Ok(Vec::new());
    }
    if !io::stdin().is_terminal() || !io::stdout().is_terminal() {
        return Err(Error::NotATerminal);
    }

    enable_raw_mode()?;
    let mut stdout = io::stdout();
    execute!(stdout, EnterAlternateScreen)?;
    let backend = CrosstermBackend::new(stdout);
    let mut terminal = Terminal::new(backend)?;

    let result = run_prompt(&mut terminal, updates);

    disable_raw_mode()?;
    execute!(terminal.backend_mut(), LeaveAlternateScreen)?;
    terminal.show_cursor()?;

    result
}

fn run_prompt(
    terminal: &mut Terminal<CrosstermBackend<Stdout>>,
    updates: &[ResolvedUpdate],
) -> Result<Vec<usize>, Error> {
    let mut selected = vec![false; updates.len()];
    let mut collapsed: HashSet<String> = HashSet::new();
    let mut cursor = 0usize;
    let mut state = ListState::default();
    state.select(Some(0));

    loop {
        let visible_rows = build_visible_rows(updates, &collapsed);
        if cursor >= visible_rows.len() {
            cursor = visible_rows.len().saturating_sub(1);
        }

        terminal
            .draw(|frame| render(frame, updates, &visible_rows, &selected, cursor, &mut state))?;

        match read_key()? {
            Key::Up => {
                if cursor > 0 {
                    cursor -= 1;
                } else {
                    cursor = visible_rows.len() - 1;
                }
                state.select(Some(cursor));
            }
            Key::Down => {
                if cursor + 1 < visible_rows.len() {
                    cursor += 1;
                } else {
                    cursor = 0;
                }
                state.select(Some(cursor));
            }
            Key::Toggle => {
                if let VisibleRow::Update { original_index } = &visible_rows[cursor] {
                    selected[*original_index] = !selected[*original_index];
                }
            }
            Key::ToggleAll => toggle_all(&mut selected),
            Key::ToggleFile => {
                let file = file_at_cursor(&visible_rows, cursor, updates);
                toggle_file(updates, &mut selected, &file);
            }
            Key::ToggleCollapse => {
                let file = file_at_cursor(&visible_rows, cursor, updates);
                if collapsed.contains(&file) {
                    collapsed.remove(&file);
                } else {
                    collapsed.insert(file);
                }
                let new_visible = build_visible_rows(updates, &collapsed);
                if cursor >= new_visible.len() {
                    cursor = new_visible.len().saturating_sub(1);
                }
                state.select(Some(cursor));
            }
            Key::Invert => invert_selected(&mut selected),
            Key::SelectNone => selected.fill(false),
            Key::Accept => {
                return Ok(selected
                    .into_iter()
                    .enumerate()
                    .filter_map(|(index, is_selected)| is_selected.then_some(index))
                    .collect());
            }
            Key::Cancel => return Err(Error::Canceled),
            Key::Ignore => {}
        }
    }
}

fn render(
    frame: &mut ratatui::Frame<'_>,
    updates: &[ResolvedUpdate],
    visible_rows: &[VisibleRow],
    selected: &[bool],
    cursor: usize,
    state: &mut ListState,
) {
    let area = frame.area();
    let sections = Layout::vertical([Constraint::Min(8), Constraint::Length(3)]).split(area);

    let items: Vec<ListItem<'_>> = visible_rows
        .iter()
        .scan(None, |last_file, row| {
            let result = match row {
                VisibleRow::FileHeader { file } => {
                    *last_file = Some(file.clone());
                    Some(render_file_header(file))
                }
                VisibleRow::Update { original_index } => {
                    let update = &updates[*original_index];
                    let file_changed = last_file.as_deref() != Some(update.file());
                    *last_file = Some(update.file().to_string());
                    Some(render_update_item(
                        update,
                        selected[*original_index],
                        file_changed,
                    ))
                }
            };
            result
        })
        .collect();

    let mut list_state = ListState::default();
    list_state.select(Some(cursor));
    *state = list_state;

    let list = List::new(items)
        .block(Block::default().borders(Borders::ALL).title("Updates"))
        .highlight_style(
            Style::default()
                .bg(Color::Rgb(45, 52, 64))
                .add_modifier(Modifier::BOLD),
        )
        .highlight_symbol("› ")
        .repeat_highlight_symbol(false)
        .scroll_padding(visible_scroll_padding(visible_rows.len()));
    frame.render_stateful_widget(list, sections[1], state);

    let footer = Paragraph::new(FOOTER)
        .block(Block::default().borders(Borders::ALL).title("Keys"))
        .wrap(Wrap { trim: true })
        .style(Style::default().fg(Color::Cyan));
    frame.render_widget(footer, sections[2]);
}

fn render_file_header(file: &str) -> ListItem<'_> {
    let mut lines = Vec::new();
    lines.push(Line::from(vec![
        Span::styled("▸ ", Style::default().fg(Color::Cyan)),
        Span::styled(
            file,
            Style::default()
                .fg(Color::Cyan)
                .add_modifier(Modifier::BOLD),
        ),
    ]));
    lines.push(Line::from(vec![
        Span::styled("  ⋯ ", Style::default().fg(Color::DarkGray)),
        Span::styled("collapsed", Style::default().fg(Color::DarkGray)),
    ]));
    ListItem::new(lines)
}

fn render_update_item(
    update: &ResolvedUpdate,
    is_selected: bool,
    file_changed: bool,
) -> ListItem<'_> {
    let marker = if is_selected { "●" } else { "○" };
    let marker_style = if is_selected {
        Style::default().fg(Color::Green)
    } else {
        Style::default().fg(Color::DarkGray)
    };
    let target_style = if update.has_sha_mismatch() || update.is_branch_ref() {
        Style::default().fg(Color::Yellow)
    } else if update.is_major_update() {
        Style::default().fg(Color::Red)
    } else {
        Style::default().fg(Color::Green)
    };
    let mut lines = Vec::new();

    if file_changed {
        lines.push(Line::from(vec![
            Span::styled("▸ ", Style::default().fg(Color::Cyan)),
            Span::styled(
                update.file(),
                Style::default()
                    .fg(Color::Cyan)
                    .add_modifier(Modifier::BOLD),
            ),
        ]));
    }

    lines.push(Line::from(vec![
        Span::styled(marker, marker_style),
        Span::raw(" "),
        Span::styled(
            &update.action,
            Style::default().add_modifier(Modifier::BOLD),
        ),
    ]));

    lines.push(Line::from(vec![
        Span::raw("    "),
        Span::styled("from ", Style::default().fg(Color::DarkGray)),
        Span::styled(&update.current, Style::default().fg(Color::Yellow)),
        if update.has_version_comment() {
            Span::styled(" (", Style::default().fg(Color::DarkGray))
        } else {
            Span::raw("")
        },
        if update.has_version_comment() {
            Span::styled(update.version_comment(), Style::default().fg(Color::Blue))
        } else {
            Span::raw("")
        },
        if update.has_version_comment() {
            Span::styled(")", Style::default().fg(Color::DarkGray))
        } else {
            Span::raw("")
        },
    ]));

    lines.push(Line::from(vec![
        Span::raw("    "),
        Span::styled("to ", Style::default().fg(Color::DarkGray)),
        Span::styled(update.display_target(), target_style),
    ]));

    let mut meta = vec![
        Span::raw("    "),
        Span::styled("job ", Style::default().fg(Color::DarkGray)),
        Span::raw(&update.job),
    ];
    if update.has_sha_mismatch() {
        meta.push(Span::styled(
            "  expected ",
            Style::default().fg(Color::DarkGray),
        ));
        meta.push(Span::styled(
            short_sha_or_full(update),
            Style::default().fg(Color::Yellow),
        ));
    }
    if update.is_branch_ref() {
        meta.push(Span::styled(
            "  (unpinned branch ref)",
            Style::default().fg(Color::Yellow),
        ));
    }
    lines.push(Line::from(meta));

    ListItem::new(lines)
}

fn visible_scroll_padding(count: usize) -> usize {
    if count <= VISIBLE_ROWS_HINT { 0 } else { 2 }
}

fn read_key() -> Result<Key, Error> {
    loop {
        let event = event::read()?;
        let Event::Key(key_event) = event else {
            continue;
        };
        if key_event.kind != KeyEventKind::Press {
            continue;
        }

        return Ok(match key_event.code {
            KeyCode::Up | KeyCode::Char('k') => Key::Up,
            KeyCode::Down | KeyCode::Char('j') => Key::Down,
            KeyCode::Enter => Key::Accept,
            KeyCode::Char(' ') | KeyCode::Char('x') => Key::Toggle,
            KeyCode::Char('a') => Key::ToggleAll,
            KeyCode::Char('f') => Key::ToggleFile,
            KeyCode::Tab => Key::ToggleCollapse,
            KeyCode::Char('i') => Key::Invert,
            KeyCode::Char('n') => Key::SelectNone,
            KeyCode::Char('q') | KeyCode::Esc => Key::Cancel,
            KeyCode::Char('c') if key_event.modifiers.contains(KeyModifiers::CONTROL) => {
                return Err(Error::Interrupted);
            }
            _ => Key::Ignore,
        });
    }
}

fn short_sha(sha: &str) -> &str {
    &sha[..sha.len().min(7)]
}

fn short_sha_or_full(update: &ResolvedUpdate) -> String {
    if !update.has_current_ref() {
        update.current.clone()
    } else {
        short_sha(update.current_ref()).to_string()
    }
}

fn invert_selected(selected: &mut [bool]) {
    for is_selected in selected {
        *is_selected = !*is_selected;
    }
}

fn toggle_all(selected: &mut [bool]) {
    let all_selected = selected.iter().all(|is_selected| *is_selected);
    selected.fill(!all_selected);
}

fn toggle_file(updates: &[ResolvedUpdate], selected: &mut [bool], file: &str) {
    let all_selected = updates
        .iter()
        .zip(selected.iter())
        .filter(|(update, _)| update.file() == file)
        .all(|(_, is_selected)| *is_selected);

    for (update, is_selected) in updates.iter().zip(selected.iter_mut()) {
        if update.file() == file {
            *is_selected = !all_selected;
        }
    }
}
